### PR TITLE
Enhanced Dockerfiles

### PR DESCRIPTION
**Pull request by brunoborges**
This PR addresses the following issues:

- No locale set by default #15 
- Perform apt/yum upgrade #13 
- Install ca-certificates, tzdata and font packages #17 
- Perform a -Xshare:dump for faster startup time #16

### CODE REVIEW

File: docker/cbld/Dockerfile.msopenjdk-11-jdk [hunk 1]
```diff
@@ -1,12 +1,29 @@
 FROM sbidprod.azurecr.io/quinault
 
-RUN apt-get update && \
-    apt-get -y install apt-transport-https && \
-    echo "deb https://packages.microsoft.com/repos/cbl-d quinault-universe main" >> /etc/apt/sources.list && \
-    apt-get update && \
-    apt-get -y install msopenjdk-11 && \
-    apt-get -y purge apt-transport-https && \
-    apt -y autoremove --purge && \
+MAINTAINER Microsoft OpenJDK Support <openjdk-support@microsoft.com>
```
Comment (d3r3kk) [inline docker/cbld/Dockerfile.msopenjdk-11-jdk:3]: Nice one.

File: docker/cbld/Dockerfile.msopenjdk-11-jdk [hunk 2]
```diff
@@ -1,12 +1,29 @@
...
+    apt-get -qq install --no-install-recommends tzdata ca-certificates fontconfig locales apt-transport-https wget && \
+    echo "en_US.UTF-8 UTF-8" >> /etc/locale.gen && \
+    locale-gen en_US.UTF-8 && \
     rm -rf /var/lib/apt/lists/*
 
-ENV JAVA_HOME=/usr/lib/jvm/msopenjdk-11-amd64
\ No newline at end of file
+ENV LANG='en_US.UTF-8' LANGUAGE='en_US:en' LC_ALL='en_US.UTF-8'
+
+RUN wget -q https://packages.microsoft.com/config/ubuntu/20.04/packages-microsoft-prod.deb -O packages-microsoft-prod.deb && \
```
Comment (d3r3kk) [inline docker/cbld/Dockerfile.msopenjdk-11-jdk:15]: Can we not run this sequence inside the previous RUN sequence? Cuts down on the number of layers IIRC to do it that way.

File: docker/ubuntu/Dockerfile.msopenjdk-17-jdk [hunk 3]
```diff
@@ -1,13 +1,29 @@
...
+
+RUN wget -q https://packages.microsoft.com/config/ubuntu/20.04/packages-microsoft-prod.deb -O packages-microsoft-prod.deb && \
     dpkg -i packages-microsoft-prod.deb && \
     apt-get -qq update && \
-    apt-get -qq -y install msopenjdk-17 && \
-    apt-get -qq -y purge apt-transport-https wget && \
-    apt -y autoremove --purge && \
+    apt-get -qq install msopenjdk-17 && \
+    apt-get -qq purge apt-transport-https wget && \
+    apt-get -qq autoremove --purge && \
```
Comment (d3r3kk) [inline docker/ubuntu/Dockerfile.msopenjdk-17-jdk:20]: Good catch.

File: docker/cbld/Dockerfile.msopenjdk-11-jdk (hunk 2 above)
Comment (d3r3kk) [reply docker/cbld/Dockerfile.msopenjdk-11-jdk:15]: (This happens in a few other places as well in this change - can we collapse them all into one for each?)

**Review by d3r3kk — COMMENTED** [review]
Just a question about combining RUN commands to reduce docker image layers, but otherwise looks good.

File: docker/cbld/Dockerfile.msopenjdk-11-jdk (hunk 2 above)
Comment (brunoborges) [reply docker/cbld/Dockerfile.msopenjdk-11-jdk:15]: Yes, it can.

File: docker/cbld/Dockerfile.msopenjdk-11-jdk [hunk 4]
```diff
@@ -1,12 +1,29 @@
 FROM sbidprod.azurecr.io/quinault
 
-RUN apt-get update && \
-    apt-get -y install apt-transport-https && \
-    echo "deb https://packages.microsoft.com/repos/cbl-d quinault-universe main" >> /etc/apt/sources.list && \
```
Comment (karianna) [inline docker/cbld/Dockerfile.msopenjdk-11-jdk:5]: Should we not retain this line?

**Review by karianna — CHANGES_REQUESTED** [review]
One Q about removing the cbl-d specific base image + I think the various sections could be commented a little e.g. Setting these vars to UTF-8 for consistent locale handling and so forth

File: docker/cbld/Dockerfile.msopenjdk-11-jdk (hunk 4 above)
Comment (brunoborges) [reply docker/cbld/Dockerfile.msopenjdk-11-jdk:5]: Yes, this should've been kept!

**Comment by d3r3kk** [comment]
Handle PR comments.

**Comment by karianna** [comment]
note on the Dockerfiles: you can actually remove the -Xshare:dump present in the JDK 17 ones.

If we think the layers are self-explanatory then that's the last fix we need.  Else would be good to add some comments in the config explaining what some of it does

**Comment by d3r3kk** [comment]
> note on the Dockerfiles: you can actually remove the -Xshare:dump present in the JDK 17 ones.
> 
> If we think the layers are self-explanatory then that's the last fix we need. Else would be good to add some comments in the config explaining what some of it does

I've removed the extra RUN statement altogether - this was to test that the java installation was working but it unnecessarily increases the container image size by ~5Mb (not much, but also has limited utility). We can move that test into a GH action instead - I have some scripts I can submit shortly that will do the trick.

**Review by karianna — APPROVED** [review]

**Comment by brunoborges** [comment]
@d3r3kk @karianna the Xshare:dump was not intended to test, but to generate meta files for faster startup time.

The extra 5mb were exactly due to such file generation. This file exists by default with JDK 17.

I'll submit a new PR to add this back.

**Pull request merged**